crates/circuit/src/dag_circuit.rs

```diff
@@ -4967,8 +4965,13 @@ impl DAGCircuit {
                 _ => Ok(false),
             }
         };
-        rustworkx_core::dag_algo::collect_runs(&self.dag, filter_fn)
-            .map(|node_iter| node_iter.map(|x| x.unwrap()))
+
+        match rustworkx_core::dag_algo::collect_runs(&self.dag, filter_fn) {
+            Some(iter) => Ok(iter.filter_map(|result| result.ok())),
+            None => Err(PyRuntimeError::new_err(
+                "Invalid DAGCircuit, cycle encountered",
+            )),
+        }
```

Comment:
```suggestion
        match rustworkx_core::dag_algo::collect_runs(&self.dag, filter_fn) {
            Some(iter) => iter.map(|result| result.unwrap()),
            None => panic!("invalid DAG: cycle(s) detected!"),
        }
```

I would recommend that we instead `panic!()` in the case that the `DAGCircuit` contains a cycle. If it does, then there's definitely a bug somewhere in one of `DAGCircuit`'s other methods that led to it corrupting itself. Such a failure is non-recoverable since we'd have no way of knowing what went wrong (if we did, then we wouldn't have introduced the bug!).

Separately, while `filter_map` is semantically equivalent to the original use of `map(|x| x.unwrap())` in this case, it's comparatively a bit misleading. The only way that the inner result can fail is if `filter_fn` defined above could return an error (which it cannot, since its `Error` type is the nonconstructible `Infallible` type). So the `filter_map` would never actually be filtering anything 🙂.